src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -970,100 +970,20 @@ protected override void ValidateElement(object element)
                         Metadata.ValidateNotNullFailure);
             }
 
-            if(_rangeKind.HasValue)
+            var o = element as PSObject;
+            if (o != null)
             {
-                // Using dynamic to handle different type of numbers such as
-                // int, double, decimal and float. 
-                dynamic val = element;
-                switch (_rangeKind)
-                {
-                    case ValidateRangeKind.Positive:
-                        if(val.CompareTo(0) <= 0)
-                        {
-                            throw new ValidationMetadataException(
-                                "ValidateRangeTooSmall",
-                                null, 
-                                Metadata.ValidateRangeSmallerThanMinRangeFailure,
-                                element.ToString(), 
-                                "1");
-                        }
-                        break;
-                    case ValidateRangeKind.NonNegative:
-                        if(val.CompareTo(0) < 0)
-                        {
-                            throw new ValidationMetadataException(
-                                "ValidateRangeTooSmall",
-                                null, 
-                                Metadata.ValidateRangeSmallerThanMinRangeFailure,
-                                element.ToString(), 
-                                '0');
-                        }
-                        break;
-                    case ValidateRangeKind.Negative:
-                        if(val.CompareTo(0) >= 0)
-                        {
-                            throw new ValidationMetadataException(
-                                "ValidateRangeTooBig",
-                                null, 
-                                Metadata.ValidateRangeGreaterThanMaxRangeFailure,
-                                element.ToString(), 
-                                "-1");
-                        }
-                        break;
-                    case ValidateRangeKind.NonPositive:
-                        if(val.CompareTo(0) > 0)
-                        {
-                            throw new ValidationMetadataException(
-                                "ValidateRangeTooBig",
-                                null, 
-                                Metadata.ValidateRangeGreaterThanMaxRangeFailure,
-                                element.ToString(), 
-                                "-1");
-                        }
-                        break;
-                }
+                element = o.BaseObject;
+            }
+            
+            if (_rangeKind.HasValue)
+            {
+                ValidatePredefinedRange(element);
```

Comment:
Comparing is the same for any range kinds (including "user defined"). I believe we should create a common method to compare elements and call it with different arguments for different ranges.